consensus/ibft/src/main/java/tech/pegasys/pantheon/consensus/ibft/statemachine/IbftController.java

```diff
@@ -127,6 +137,17 @@ private void handleMessage(final MessageData messageData) {
     }
   }
 
+  private <P extends Payload> void processMessage(
+      final Message message,
+      final SignedData<P> signedPayload,
+      final BiConsumer<SignedData<P>, Address> gossipMessage,
+      final Consumer<SignedData<P>> handleMessage) {
+    if (processMessage(signedPayload, message)) {
+      gossipMessage.accept(signedPayload, message.getConnection().getPeer().getAddress());
```

Comment:
If this is reworked as per:
  private <P extends Payload> void processMessage(
      final Message message,
      final SignedData<P> signedPayload,
      final Consumer<SignedData<P>> handleMessage) {
    if (processMessage(signedPayload, message)) {
      gossiper.gossipMessage(signedPayload.getSignature(),
          message.getData(),
          signedPayload.getSender(),
          signedPayload.getSender(), message.getConnection().getPeer().getAddress());
      handleMessage.accept(signedPayload);
    }
  }

The gossip Biconsumer can be removed, and the msg specific handlers in Gossiper can be removed.

Reply:
I don't think this is an improvement. There's nothing wrong with using a Biconsumer. The suggested version makes using the Gossiper require additional co-ordination and moves the responsibility for identifying the message and it's signer to the invoker which is worse encapsulation.